source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,927 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+#include <deal.II/fe/fe_enriched.h>
+#ifdef DEAL_II_WITH_CXX11
+
+#include <deal.II/fe/fe_tools.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  /**
+   * Auxiliary function to create multiplicity vector from input enrichment functions.
+   */
+  template <typename T>
+  std::vector<unsigned int>
+  build_multiplicities(const std::vector<std::vector<T > > &functions )
+  {
+    std::vector<unsigned int> multiplicities;
+    multiplicities.push_back(1); // the first one is non-enriched FE
+    for (unsigned int i = 0; i < functions.size(); i++)
+      multiplicities.push_back(functions[i].size());
+
+    return multiplicities;
+  }
+
+
+  /**
+   * Auxiliary function to build FiniteElement's vector
+   */
+  template <int dim, int spacedim>
+  std::vector< const FiniteElement< dim, spacedim > * >
+  build_fes(const FiniteElement<dim,spacedim> *fe_base,
+            const std::vector<const FiniteElement<dim,spacedim> * > &fe_enriched)
+  {
+    std::vector< const FiniteElement< dim, spacedim > * > fes;
+    fes.push_back(fe_base);
+    for (unsigned int i = 0; i < fe_enriched.size(); i++)
+      fes.push_back(fe_enriched[i]);
+
+    return fes;
+  }
+
+
+  /**
+   * Auxiliary function which check consistency of the input parameters.
+   * Returns true if everything is ok.
+   */
+  template <int dim, int spacedim>
+  bool
+  consistency_check (const std::vector< const FiniteElement< dim, spacedim > * > &fes,
+                     const std::vector< unsigned int > &multiplicities,
+                     const std::vector<std::vector<std::function<const Function<spacedim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > &functions)
```

Comment:
It would be lovely if this function could check that the enrichment functions are really all scalar, because that's really the only case you support right now. But you can't do that here because you no longer have access to the `Function` objects without picking any particular cell.

I'm going to comment on this again below in `fill_fe_values`.